GUI: Port to Qt5, implement CMake build, port to Linux, make Draco optional

lubosz opened the pull request:
During my vacation in December I took some time to make the Compressonator GUI build and run on Linux. This is a rather long patch set, I hope you will have time to review it. I am ready to improve this merge request if you see any problems.

The patch set should not break anything on other platforms, but I did not build the branch on MSVC or macOS to test it.

This merge request contains the following changes:
1. Port to Qt5, especially the local dependency `QPropertyPages`, which is not available in Qt5 since upstream is dead.
2. Implementation of a CMake build for the GUI and required dependencies
3. Resolution of platform specific issues like backslashes, case sensitivity in build and runtime and sleep functions. The patches should build and run on all platforms.
4. `#ifdef`ing of Windows platform specific code using the available `_WIN32` define, as seen in the CLI.
5. Optional compilation of Draco when it is not found as an external dependency in CMake. If you decide to use CMake on Windows using Draco as an external dependency as well, the Draco code base could be removed from the Compressonator repository. If you decide to keep it, I would encourage you to use it as a git submodule for maintainability and usage of the Draco CMake files.
6. Implementing missing enums and removing uncalled code that required missing headers.

Details can be found in the respective commit messages.

This patch set depends on my prior merge request https://github.com/GPUOpen-Tools/Compressonator/pull/86

NPCompress commented:
@lubosz. An excellent contribution to this repository and thank you for the work to make the required changes to the CMake files and GUI code, the changes will be put through our build system and pulled in after verification. The executable build will then be added to the release for Linux GUI support.

lubosz commented:
I added install targets, so the project can be install and packaged more easily and resolved some issues with library orders at link time during creating the Arch package.

My Compressonator package is now available in the Arch Linux user repository:
https://aur.archlinux.org/packages/compressonator-git

axredneck commented:
```
CMake Error in Applications/CompressonatorCLI/CMakeLists.txt:
  Imported target "Boost::system" includes non-existent path

    "/include"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
```
Arch Linux, your PKGBUILD

lubosz commented:
@axredneck Thanks for testing this.
I would encourage you to post issues related to the PKGBUILD on the AUR, to keep this discussion related to the patch set itself.

Apparently your CMake Boost configuration is broken. Do you have any custom CMake / boost related packages installed, that are not from the Arch main repo? E.g. `pacman -Qm`. Your CMake configuration is looking for an `/include` directory, so it's apparently missing the `/usr` prefix correctly supplied in the PKGBUILD. Do you have any custom config in `makepkg.cong`?

To make sure that the problem is not on my end I will build the package in a fresh Arch docker and see if it requires any changes.

I also pushed one more patch to remove hardcoded include paths from the CLI build, this shouldn't really have an impact on your problem, but you never know.

lubosz commented:
@axredneck 
I could reproduce your problem in a similar way with a minimal Arch Linux docker installation.
The problem was on my side. It should be resolved now.

There were several unfortunate issues for you getting this problem:
1. The Boost dependency was wrongfully not set to `REQUIRED` in CMake, the most recent patch in this set resolved this.
2. The Arch Linux package was missing the `boost` build dependency, which was also not detected by `namcap`, which only picked up the runtime dependency `boost-libs`.

axredneck commented:
@lubosz i still get the same error

lubosz commented:
> @lubosz i still get the same error

@axredneck 
Try cleaning your build directory. The AUR package currently does an in-source build, should probably change that. E.g. `rm ~/.cache/yay/compressonator-git` if you use yay.

I also was getting a different boost include error than you on the minimal docker install (missing header during build, your problem is during configuration time).

So if cleaning the AUR build cache does not resolve your problem, look into the steps above to see if you have any custom shenanigans going on with Boost or makepkg.

The package builds fine now in a fresh Arch install after I added Boost to the `makedepends` and removed the unused glut includes.

NPCompress commented:
@lubosz,  Test status update: The application code changes in this pull request breaks the current behavior of the Windows GUI image views. The code changes will have to be reviewed further and updated to fix  this, before the merge pull request is added to the master branch.

lubosz commented:
@NPCompress thanks for testing this.
What comes to my mind that could actually change behavior are the patches related to Qt5 or the ones where I cast function pointers to void. All in all a `git bisect` could be handy for detection of the faulty patches.
I will set up a Windows build on the week end to investigate this.

lubosz commented:
> @lubosz, Test status update: The application code changes in this pull request breaks the current behavior of the Windows GUI image views.
@NPCompress Could you please point out which behavior of the GUI was changed in Windows, so I can fix it?

NPCompress commented:
@lubosz, I did not get the time to review your request for more info on the windows image issues when merging your changes, will do so next week.

NPCompress commented on Compressonator/Applications/CompressonatorGUI/QPropertyPages/qtvariantproperty.cpp:
```diff
@@ -1649,73 +1649,73 @@ void QtVariantPropertyManager::setValue(QtProperty *property, const QVariant &va
         return;
     } else if (QtDoublePropertyManager *doubleManager = qobject_cast<QtDoublePropertyManager *>(manager)) {
-        doubleManager->setValue(internProp, qVariantValue<double>(val));
+        doubleManager->setValue(internProp, val.value<int>());
```
This should be double

NPCompress commented on Compressonator/Applications/CompressonatorGUI/Components/acImageView.cpp:
```diff
@@ -1330,14 +1330,15 @@ void acImageView::centerImage()
-    m_imageItem_Processed->translate(dx, dy);
+    QTransform translation;
+    translation.translate(dx, dy);
+    m_imageItem_Processed->setTransform(translation);
```
change this to m_imageItem_Processed->setTransform(translation,true);

NPCompress commented on Compressonator/Applications/CompressonatorGUI/Components/acImageView.cpp:
```diff
@@ -1330,14 +1330,15 @@ void acImageView::centerImage()
     {
         m_imageItem_Original->setTransformOriginPoint(0, 0);
-        m_imageItem_Original->translate(dx, dy);
+        m_imageItem_Original->setTransform(translation);
```
change this to m_imageItem_Original->setTransform(translation,true);

lubosz commented:
> What version of Qt are you using?
I was using the current release version `5.14.1`.
> When adding the transform for v5.7.1, the image views are no longer positioned correctly, they are shifted with the top edge of image been at the center of the view.
> 
> To fix this change to your code is to add true to the setTrandform calls
> 
> setTransform(translation**,true**);
Thanks for the review, I will update the merge request.

lubosz commented:
I updated and rebased the branch with the said changes.

lubosz commented:
Awesome, thanks. Looking forward to contribute more.

NPCompress commented:
@lubosz Hi a new version of Windows GUI was posted recently to support OpenCL and DirectX, The QT updates for Linux were not reviewed for the build of OpenCL inclusion. let us know if there are any changes you made for building the Linux GUI so it can be made available on the next release.
